docs/architecture/deep-dives/authorization.md

```diff
@@ -0,0 +1,247 @@
+a `NotFoundError` if the check fails.
+
+### Create
+
+Instantiate the object you want to save, then pass it to `AuthorizationService`.
+
+```cs
+var cipher = cipherRequestModel.ToCipher();
+await _authorizationService.AuthorizeOrThrowAsync(User, cipher, CipherOperations.Create);
+
+await _cipherRepository.Create(cipher);
+```
+
+### Read
+
+Read the object from the database, then pass it to `AuthorizationService`.
+
+```cs
+var cipher = _cipherRepository.GetByIdAsync(id);
+await _authorizationService.AuthorizeOrThrowAsync(User, cipher, CipherOperations.Read);
+
+return new CipherResponseModel(cipher);
+```
+
+### Update
+
+Read the **unedited** object from the database, then pass it to `AuthorizationService`.
+
+```cs
+var cipher = _cipherRepository.GetByIdAsync(id);
+await _authorizationService.AuthorizeOrThrowAsync(User, cipher, CipherOperations.Update);
+
+// Only update the cipher after the authorization check has passed
+cipher.Name = cipherRequest.Name;
+await _cipherRepository.UpdateAsync(cipher);
+```
+
+:::danger
+
+Do **not** use the request object as the authorization input. It has been provided by the user and
+is not a trusted source of authorization.
+
+:::
+
+### Delete
+
+Read the object from the database, then pass it to `AuthorizationService`.
+
+```cs
+var cipher = _cipherRepository.GetByIdAsync(id);
+await _authorizationService.AuthorizeOrThrowAsync(User, cipher, CipherOperations.Delete);
+
+await _cipherRepository.DeleteAsync(cipher);
+```
+
+### Bulk reads
+
+Some queries return all resources of a type within a particular scope. For example, rather than
+returning a specific cipher, return all ciphers for an organization.
+
+In this example, the `CurrentContextOrganization` object (representing the organization) becomes the
+resource, and the operation describes the scope of the read. This would require a separate handler
+to be defined for this combination of resource and operation.
+
+```cs
+var organization = _currentContext.GetOrganization(orgId);
+await _authorizationService.AuthorizeOrThrowAsync(User, organization, CipherOperations.ReadAllForOrganization);
+
+var result = await _cipherRepository.ReadManyByOrganizationId(orgId);
+```
+
+Sometimes the database query itself is scoped to the user, such that no additional authorization
+check is required or even possible. If this is not obvious from the context, note this in a comment:
+
+```cs
+// Note: this database call only returns the user's ciphers - no authorization check needed
+var result = await _cipherRepository.ReadManyByUserId(userId);
+```
+
+## Guidelines
+
+### CQRS
+
+Authorization checks (i.e. the call to `IAuthorizationService`) should be contained in
+[command and query classes](http://localhost:3000/architecture/server/#cqrs-adr-0008).
+
+This is the simplest way to ensure that authorization is always checked, and ensures that the
+authorization check stays in step with what the query or command actually does.
```

Comment:
@justindbaur this is what we discussed on the call, but I realised this means we must pass in the `ClaimsPrincipal` to almost all of our commands/queries, or make it available via `CurrentContext` for injection. Do you think that's an OK path to go down? It feels like it's distorting the shape of our commands. On the other hand, we often need to pass in a `UserId` anyway.

I also remember you were concerned about `HttpContext` disappearing - not sure if that's an issue here as the `ClaimsPrincipal` is passed through the core layer.

Reply:
An alternative that I discussed with @MGibson1 is that commands only return a list of resources & authorization requirements, which the controller can then use to call `AuthorizationService`. The command doesn't call `AuthorizationService` itself. That prevents the `ClaimsPrincipal` from being passed into `Core`, and also makes the authz check very visible in the api layer.

Incidentally, that also means that you can execute the command without an authz check. This can pose a risk, but it's sometimes required where the command is being called as a side-effect and you want it to always execute with fixed arguments even if the user doesn't have permission to call the command otherwise. (e.g. verifying a domain turns on the single org policy, even if the user doesn't have permission to modify policies.)

Reply:
An alternative that I discussed with @MGibson1 is that commands only return a list of resources & authorization requirements, which the controller can then pass to AuthorizationService. (e.g. `command.GetAuthorizationRequirements`.) The command doesn't call AuthorizationService itself. That prevents the ClaimsPrincipal from being passed into Core, and also makes the authz check very visible in the api layer. I am very much in favour of both those.

Another benefit is that you can execute the command without an authz check. This can pose a risk, but it's sometimes required where the command is being called as a side-effect and you want it to always execute with fixed arguments regardless of permissions. (e.g. verifying a domain turns on the single org policy, even if the user doesn't have permission to modify policies.)

Reply:
Having the `ClaimsPrincipal` in our context seems to be the smoothest to me, so that commands can encapsulate a lot of this. Are there concerns with that?

Reply:
After thinking about it further and talking to Justin, I don't see any problems here.